### PR TITLE
Implemented initial POC

### LINKED ISSUE #1: Obtain usage data for bare metal and generate initial invoice
As part of nerc-project/operations#831, we want to have a way to pull usage for bare metal (BM), and turn them into invoices that we can process with `CCI-MOC/invoicing`, similar to `CCI-MOC/openshift-usage-scripts` and `CCI-MOC/openstack-billing-from-db`.

To do:
- [x] Determine storage locations for the BM invoices (likely S3?)

@joachimweyl @knikolla @naved001 Let me know if there are more things that need to be considered.

### PR DESCRIPTION
**Pull request by QuanMPhm**
Closes #1. I submitted this as a draft since I wanted everyone's feedback on how the repo is structured before fully implementing every nitty-gritty detail. This repo is mostly based on the [openstack billing repo](https://github.com/CCI-MOC/openstack-billing-from-db), with a few small differences. Below is some more explanation of some of the structural choices I made:

- For now, while I am deciding how to BM usage data should be fetched, the entrypoint script `main.py` expects the usage data to provided through the CLI
- The script does not contain a Coldfront client, under the assumption that CCI-MOC/invoicing#159 will be merged before this.
- Unlike the [various data models](https://github.com/CCI-MOC/openstack-billing-from-db/blob/5b980749a8bd4d95834ace675b51d87f02fb4abb/src/openstack_billing_db/billing.py#L30-L34) in the openstack billing, I tried to minimize where the list of known SUs are hardcoded. This is too reduce future maintenance burden, and is related to this [discussion](https://github.com/CCI-MOC/bare-metal-usage-scripts/issues/1#issuecomment-2740670320)
- `[Pydantic](https://docs.pydantic.dev/latest/)` is used validate the input BM usage data

Let me know what you guys think!

### CODE REVIEW

File: src/bare_metal_billing/billing.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+from dataclasses import dataclass
+from decimal import Decimal
+
+from bare_metal_billing.models import ProjectUsage, BMUsageData
+
+
+@dataclass
```
Comment (larsks) [inline src/bare_metal_billing/billing.py:7]: Why not use pydantic here? Especially since you're *already* using it in `models.py`. Using `dataclass` doesn't get you any sort of validation. Compare this example, which uses the `dataclass` definition:

```
>>> SURates(su_rates=None)
SURates(su_rates=None)
```

With the following code, which uses a Pydantic-based model:

```
>>> SURatesPydantic(su_rates=None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lars/tmp/python/.venv/lib/python3.12/site-packages/pydantic/main.py", line 214, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for SURatesPydantic
su_rates
  Input should be a valid dictionary [type=dict_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.10/v/dict_type
```

To make the type annotations meaningful if you're using `dataclass`, you need to add a type checker to your CI scripts (e.g., mypi or basedpyright or similar).

File: src/bare_metal_billing/utils.py
Comment (larsks) [inline src/bare_metal_billing/utils.py:1]: Please don't create modules called `utils`, because they end up becoming dumping grounds. Since `parse_time_from_string` is only called from the `main` module, move it into `main.py`.

File: src/bare_metal_billing/billing.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+class InvoiceWriter:
+    HEADERS = [
+        "Invoice Month",
+        "Project - Allocation",
+        "Project - Allocation ID",
+        "Invoice Email",
+        "Invoice Address",
+        "Institution",
+        "Institution - Specific Code",
```
Comment (larsks) [inline src/bare_metal_billing/billing.py:42]: This should probably be `Institution-specific Code`, because `Institution-specific` is a compound word.

File: src/bare_metal_billing/models.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    if isinstance(v, str):
+        return datetime.datetime.fromisoformat(v)
+    return v
+
+
+def parse_bm_accelerator(v: list) -> int:
+    return len(v)
+
+
+PositiveInt = Annotated[int, pydantic.Field(gt=0)]
```
Comment (larsks) [inline src/bare_metal_billing/models.py:19]: This doesn't appear to be used anywhere.

File: src/bare_metal_billing/models.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+DateField = Annotated[datetime.datetime, pydantic.BeforeValidator(parse_date)]
+
+
+class BMResourceProperties(pydantic.BaseModel):
+    cpu_count: Annotated[int, pydantic.Field(alias="cpus", gt=-1)] = 0
+    ram: Annotated[int, pydantic.Field(alias="memory_mb", gt=-1)] = 0
+    gpu_count: Annotated[
+        int,
+        pydantic.Field(alias="accelerators", gt=-1),
+        pydantic.BeforeValidator(lambda v: len(v)),
```
Comment (larsks) [inline src/bare_metal_billing/models.py:29]: What is this validator doing? `gpu_count` is an `int` value; calling `len` on it doesn't make any sense. This code results in the error:

```
>>> x = BMResourceProperties(accelerators=0)
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    x = BMResourceProperties(accelerators=0)
  File "/usr/lib/python3.13/site-packages/pydantic/main.py", line 214, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
  File "/home/lars/redhat/moc/bare-metal-usage-scripts/src/bare_metal_billing/models.py", line 29, in <lambda>
    pydantic.BeforeValidator(lambda v: len(v)),
                                       ~~~^^^
TypeError: object of type 'int' has no len()
```

It looks like you don't have tests for your validators.

File: src/bare_metal_billing/models.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+def parse_bm_accelerator(v: list) -> int:
+    return len(v)
+
+
+PositiveInt = Annotated[int, pydantic.Field(gt=0)]
+DateField = Annotated[datetime.datetime, pydantic.BeforeValidator(parse_date)]
+
+
+class BMResourceProperties(pydantic.BaseModel):
+    cpu_count: Annotated[int, pydantic.Field(alias="cpus", gt=-1)] = 0
```
Comment (larsks) [inline src/bare_metal_billing/models.py:24]: For all of the instance in which you're using `typing.Annotated` with `pydantic.Field`, you could write instead:

```
cpu_count: int = pydantic.Field(alias='cpus', default=0, gt=-1)
```

I guess I don't have a strong preference here, but I wanted to make sure you were aware that using `Annotated` was not required in these situations.

Comment (larsks) [reply src/bare_metal_billing/models.py:24]: What are all the field aliases for here?  This means that you *set* a value like this:

```
>>> x = BMResourceProperties(cpus=2)
```

But then access the value like this:

```
>>> x.cpu_count
2
```

Why not just drop the alias and pick a field name that you'll both when getting and setting the value?

File: src/bare_metal_billing/models.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        if data.expire_time and data.expire_time < data.start_time:
+            raise ValueError("expire_time must be after start_time")
+        return data
+
+
+class BMUsageData(pydantic.RootModel):
+    root: list[BMNodeUsage]
+
+
+@dataclass
```
Comment (larsks) [inline src/bare_metal_billing/models.py:54]: Why are we mixing in a `dataclass` in a module in which we're already using `pydantic` models?

**Review by larsks — CHANGES_REQUESTED** [review]

File: src/bare_metal_billing/billing.py (hunk 2 above)
Comment (QuanMPhm) [reply src/bare_metal_billing/billing.py:42]: @larsks The string `"Institution - Specific Code"` was chosen as part of our billing specification, and so it [appears](https://github.com/CCI-MOC/openshift-usage-scripts/blob/fa8e79a298d301663826c32bc4c7d2aa05ee1d61/openshift_metrics/utils.py#L132) in our [various](https://github.com/CCI-MOC/openstack-billing-from-db/blob/5b980749a8bd4d95834ace675b51d87f02fb4abb/src/openstack_billing_db/billing.py#L197) billing repo. I don't mind turning it into `Institution-specific`, but that seems like a policy question beyond me

Comment (naved001) [reply src/bare_metal_billing/billing.py:42]: @larsks what you are saying is correct and makes sense but this was how were asked to generate the report and I guess we didn't notice it then. We should probably change it in a different PR because we'll need to update all the other reports (openstack, openshift, storage) as well. But before that, I want to ask @joachimweyl if this change is acceptable.

That also means that either the code will need to work with both of these column names or we'll need to re-use a specific version of our billing code if we needed to regenerate invoices from the past.

Comment (joachimweyl) [reply src/bare_metal_billing/billing.py:42]: I suggested this when we were creating the requirements and it was vetoed. Let's leave it as requested by the requirements.

File: src/bare_metal_billing/billing.py (hunk 1 above)
Comment (QuanMPhm) [reply src/bare_metal_billing/billing.py:7]: @larsks Thank you for pointing this out. This was something I asked myself too while writing this. Isn't `pydantic` just objectively better than `dataclasses`? It just provides more useful and convenient features. I hesitated to turn everything into a Pydantic model since I saw that [other](https://github.com/CCI-MOC/openstack-billing-from-db/blob/5b980749a8bd4d95834ace675b51d87f02fb4abb/src/openstack_billing_db/model.py#L14) repos were using `dataclasses`, and inferred that the logic was:
- If the input data is something external to the code (like the BM usage data) then we should use `Pydantic`. If it is data within the code, using `dataclasses` is fine.

If you believe it makes sense to use `pydantic` everywhere, I am happy to do so.

File: src/bare_metal_billing/models.py (hunk 5 above)
Comment (QuanMPhm) [reply src/bare_metal_billing/models.py:24]: @larsks I just wanted conformity between `cpu_count` and `gpu_count`. I thought the names were more explicit

File: src/bare_metal_billing/models.py (hunk 4 above)
Comment (QuanMPhm) [reply src/bare_metal_billing/models.py:29]: @larsks In the BM usage data (you can find an example data [here](https://github.com/CCI-MOC/bare-metal-usage-scripts/blob/1bebd16a251d6f58d977ab333e1f76d944f18f70/src/bare_metal_billing/tests/test_models.py#L36-L88)), the number of GPUs a node has is not explicitly given. Instead, the GPU count is based on the number of entries in the `Resource Properties.accelerators` field, hence `lambda v: len(v)`

Comment (larsks) [reply src/bare_metal_billing/models.py:29]: This is worth documenting with a comment, and I think you need to update the validator so that it also accepts integers. You've declared a model that says, "`gpu_count` is an integer", but if you try initializing it that way, you get a `TypeError` exception. That violates the "[principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)".

File: src/bare_metal_billing/models.py (hunk 5 above)
Comment (larsks) [reply src/bare_metal_billing/models.py:24]: I think the alias is confusing, but I don't feel very strongly about it. Go with your preference.

File: src/bare_metal_billing/models.py (hunk 3 above)
Comment (QuanMPhm) [reply src/bare_metal_billing/models.py:19]: Thank you. I forgot to used this in my model :P

File: src/bare_metal_billing/models.py (hunk 6 above)
Comment (QuanMPhm) [reply src/bare_metal_billing/models.py:54]: I have turned `ProjectUsage` into a `pydantic` module

File: src/bare_metal_billing/billing.py (hunk 1 above)
Comment (QuanMPhm) [reply src/bare_metal_billing/billing.py:7]: `SURates` and `SUInfo` have been moved to `model.py` and turned into `pydantic` modules

**Comment by QuanMPhm** [comment]
I have mostly completed my initial POC, with a few lingering questions after testing my scripts on the test billing data:
@joachimweyl @knikolla @naved001 
- During testing, I saw that some leases had starting times that were AFTER their expiry times. Some leases also had missing resource information (i.e no info on CPUs or RAM). Most of these seem to be test leases that are likely not billable, but I wanted your opinion on whether I should automatically filter these out of the (unprocessed) invoice from this script, or wait for `CCI-MOC/invoicing` to handle filtering invalid projects. Should I raise errors? Warnings?
- More so for @tzumainn. Do you know if it is possible for a leased node which shouldn't have GPUs (like FC430) to still show that it does have GPUs? Could a non-GPU node still use GPUs? This impacts how I will write my test cases

Please let me know if you guys have any questions about the code or edge cases.

**Comment by tzumainn** [comment]
>     * More so for @tzumainn. Do you know if it is possible for a leased node which shouldn't have GPUs (like FC430) to still show that it does have GPUs? Could a non-GPU node still use GPUs? This impacts how I will write my test cases
> 

Nope! This shouldn't be possible.

File: bare_metal_billing/tests/test_billing.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+            self.assertEqual(
+                billing._get_su_quantity(
+                    test_bm_usage_data.root[i], self.TEST_SU_INFO, test_su_types[i]
+                ),
+                answer_su_quantities[i],
+            )
+
+    def test_get_su_quantity_fc430_with_gpu(self):
+        """What happens if a SU without GPU (i.e like FC430) has GPUs?"""
+        # TODO: Is this even possible? Should we have a test case for this? Warnings?
```
Comment (QuanMPhm) [inline bare_metal_billing/tests/test_billing.py:227]: @knikolla @naved001 [Mainn already confirmed](https://github.com/CCI-MOC/bare-metal-usage-scripts/pull/3#issuecomment-2749079768) that this shouldn't be possible. Should I still keep this test case?

File: bare_metal_billing/main.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            "BM FC830": args.rate_fc830_su,
+            "BM GPUA100SXM4": args.rate_gpu_a100sxm4_su,
+            "BM GPUH100": args.rate_gpu_h100_su,
+        }
+        for su_name in billing.BM_SU_LIST:
+            su_info_dict[su_name] = {}
+            for su_resource in billing.SU_RESOURCE_LIST:
+                su_info_dict[su_name][su_resource] = Decimal(1)
+
+    # TODO What are reasonable values for UNKNOWN SUs?
```
Comment (QuanMPhm) [inline bare_metal_billing/main.py:138]: @knikolla @naved001 @joachimweyl

Comment (naved001) [reply bare_metal_billing/main.py:138]: If I am unable to determine the SU then for an unknown SU it's 0. My rationale is that we must be sure what the SU is if we want to charge someone.

Ultimately, in the final report we will should see how many hours of unknown SU a project consumed and investigate the cause of it.

File: bare_metal_billing/tests/test_billing.py (hunk 7 above)
Comment (naved001) [reply bare_metal_billing/tests/test_billing.py:227]: Does the code look at the inspection data of the node to see if has GPUs or not?

File: bare_metal_billing/billing.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+]
+
+RESOURCE_CLASS_2_SU_MAPPING = {
+    "lenovo-sd650nv2-a100": "BM GPUA100SXM4",
+    "sd650nv2": "BM GPUA100SXM4",
+    "fc430": "BM FC430",
+    "fc830": "BM FC830",
+}
+
+SU_RESOURCE_LIST = ["vCPUs", "RAM", "vGPUs"]
```
Comment (naved001) [inline bare_metal_billing/billing.py:23]: Do we call it "vCPUs" everywhere of just "vCPU"? 

"vGPUs" is definitely incorrect, these are full physical GPUs. I don't remember what we called them in the nerc-rates code but we should probably be consistent. So, I guess either "GPUs" or "GPU" whatever we use for other repos.

File: bare_metal_billing/billing.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    cpu_count = lease_info.resource_properties.cpu_count
+    ram = lease_info.resource_properties.ram
+    gpu_count = lease_info.resource_properties.gpu_count
+
+    cpu_quantity = 0 if cpu_count == 0 else cpu_count / lease_su_info["vCPUs"]
+    ram_quantity = 0 if ram == 0 else ram / lease_su_info["RAM"]
+    gpu_quantity = 0 if gpu_count == 0 else gpu_count / lease_su_info["vGPUs"]
+
+    su_quantity = math.ceil(max(cpu_quantity, ram_quantity, gpu_quantity))
+    # TODO: What should do we for SUs with 0 SU? Meaning missing cpu and ram data, likely invalid data?
```
Comment (naved001) [inline bare_metal_billing/billing.py:95]: In the openshift world I went with setting it to 0 SUs of Unkown SU type because lots of pods continue to run without requesting any CPU and/or RAM (technically that means they don't have any guarantees).

In this case, we could do the same i.e. assign 0 SUs of Unkown SU type and log an error and continue generating the report. Or do what you are currently doing and just stop any further processing until we figure out what node had invalid data.

Comment (QuanMPhm) [reply bare_metal_billing/billing.py:95]: @naved001 @knikolla @joachimweyl Following what Naved has done in Openshift billing, if a project has 0 SUs, I will set it as having 0 SUs, and therefore $0 charged as cost. An error will be logged, but the report will still be generated.

File: bare_metal_billing/models.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+PositiveDecimalField = Annotated[Decimal, pydantic.Field(gt=0)]
+PositiveIntField = Annotated[int, pydantic.Field(gt=-1)]
+DateField = Annotated[datetime.datetime, pydantic.BeforeValidator(parse_date)]
+
+
+class BMResourceProperties(pydantic.BaseModel):
+    cpu_count: Annotated[PositiveIntField, pydantic.Field(alias="cpus")] = 0
+    ram: Annotated[PositiveIntField, pydantic.Field(alias="memory_mb")] = 0
+    gpu_count: Annotated[
```
Comment (naved001) [inline bare_metal_billing/models.py:36]: in fact, why do we even care about how many cpu or ram a node has? Aren't we directly billing based on the resourceclass?

File: bare_metal_billing/models.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        return len(v)
+    return v
+
+
+PositiveDecimalField = Annotated[Decimal, pydantic.Field(gt=0)]
+PositiveIntField = Annotated[int, pydantic.Field(gt=-1)]
+DateField = Annotated[datetime.datetime, pydantic.BeforeValidator(parse_date)]
+
+
+class BMResourceProperties(pydantic.BaseModel):
```
[the remark this answers is not in the document]
Comment (QuanMPhm) [reply bare_metal_billing/models.py:33]: From today's discussion, we agreed to bill based on the su type (i.e `fc430`) of the node, instead of going through the trouble of inspect node information and fetching resource info from them.

**Comment by QuanMPhm** [comment]
@naved001 @knikolla @joachimweyl Billing for BM is now solely determined by the node's resource type and its run time from the start time to expiry time. I've also added time clamping. Test cases have been updated. This leaves me with one last problem:
- During testing, I saw that some leases had starting times that were AFTER their expiry times. Some leases also had missing resource information (i.e no info on CPUs or RAM). Most of these seem to be test leases that are likely not billable, but I wanted your opinion on whether I should automatically filter these out of the invoice created by this script, or wait for `CCI-MOC/invoicing` to handle filtering invalid projects. Should I raise errors? Warnings?

**Comment by QuanMPhm** [comment]
@knikolla @naved001 @larsks This is now ready for review

File: bare_metal_billing/billing.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+    def write_csv(self):
+        csv_rows = [self.HEADERS]
+        for project_invoice in self.project_invoices:
+            for su_type, su_hour in project_invoice.su_hours.items():
+                csv_rows.append(
+                    (
+                        self.invoice_month,
+                        project_invoice.project_name,
+                        "ESI Bare Metal",
```
[the remark this answers is not in the document]
Comment (QuanMPhm) [reply bare_metal_billing/billing.py:56]: In the original [issue](https://github.com/nerc-project/operations/issues/831#issue-2684177015) to automate BM billing, one of the requirements was:
> Project - Allocation ID becomes "ESI Bare Metal"

Now that I think about it, are we keeping track of PI or allocation info at all for BM in Coldfront? This would mess with the Coldfront processing in `invoicing`, so I'll need to make a small adjustment to `invoicing`.

[the remark this answers is not in the document]
Comment (knikolla) [reply bare_metal_billing/billing.py:57]: The `Project - Allocation ID` field is meant to contain the value that we will search in ColdFront to match with a ColdFront allocation in the Processor.

[the remark this answers is not in the document]
Comment (QuanMPhm) [reply bare_metal_billing/billing.py:57]: Ah yes, you are absolutely right

File: bare_metal_billing/billing.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        time = min_time
+    if time > max_time:
+        time = max_time
+    return time
+
+
+def get_project_invoices(
+    bm_usage_data: models.BMUsageData, start_time: datetime, end_time: datetime
+) -> list[models.ProjectUsage]:
+    project_usage_dict = {}
```
[the remark this answers is not in the document]
Comment (QuanMPhm) [reply bare_metal_billing/billing.py:105]: I've used the pre-existing `ProjectUsage` model, and added a `add_usage` utility to the model

**Review by knikolla — COMMENTED** [review]
Seems good at a first pass, will do a second pass later today after lunch.

File: bare_metal_billing/tests/test_billing.py [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+class BillingTestBase(TestCase):
+    def _get_project_invoice_list(
+        self,
+        project_names,
+        su_hours=None,
+    ):
+        if not su_hours:
+            su_hours = [{}] * len(project_names)
+        project_invoice_list = []
+        for i in range(len(project_names)):
```
[the remark this answers is not in the document]
Comment (QuanMPhm) [reply bare_metal_billing/tests/test_billing.py:21]: In that loop, I'm indexing over both `project_names` and `su_hours`

File: bare_metal_billing/billing.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+
+BM_SU_LIST = [
+    "BM FC430",
+    "BM FC830",
+    "BM GPUA100SXM4",
+    "BM GPUH100",
+]
+
+RESOURCE_CLASS_2_SU_MAPPING = {
```
Comment (tzumainn) [inline bare_metal_billing/billing.py:16]: should this have the H100 mapping?

**Review by tzumainn — COMMENTED** [review]
One inline question; looks good otherwise!

File: bare_metal_billing/billing.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    "lenovo-sd665nv3-h100": "BM GPUH100",
+    "lenovo-sd650nv2-a100": "BM GPUA100SXM4",
+    "sd650nv2": "BM GPUA100SXM4",
+    "fc430": "BM FC430",
+    "fc830": "BM FC830",
+}
```
Comment (tzumainn) [inline bare_metal_billing/billing.py:22]: There are two other resource classes that might be relevant:

`r740xd_amdmi210`: node with an AMD Aldebaran/MI200 [Instinct MI210] GPU; there's only one, and it's leased to Triton
`fc830-nvme`: fc830 configured with nvme storage; there are 18, and all are leased to the storage-research project

Comment (QuanMPhm) [reply bare_metal_billing/billing.py:22]: @joachimweyl Can you confirm how these should be handled in billing?

Comment (naved001) [reply bare_metal_billing/billing.py:22]: @QuanMPhm Should this be read from a config file (json, yaml) - BM_SU_LIST and the mapping?

Comment (joachimweyl) [reply bare_metal_billing/billing.py:22]: for now ignore these. We do not have a pricing for them. I will keep a list of these and bring them up with the admin team to discuss how to bill for them eventually. I guess set pricing to 0 until further notice so they still exist in our records.

Comment (tzumainn) [reply bare_metal_billing/billing.py:22]: Thanks for confirming all this!

**Review by tzumainn — APPROVED** [review]

File: bare_metal_billing/models.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+class BMUsageData(pydantic.RootModel):
+    root: list[BMNodeUsage]
+
+    @pydantic.model_validator(mode="after")
+    def validate_expire_time(self):
+        validated_list = []
+        for node_usage in self.root:
+            if node_usage.expire_time:
+                if node_usage.expire_time < node_usage.start_time:
+                    continue
```
Comment (naved001) [inline bare_metal_billing/models.py:38]: So, we just ignore any lease info where the expire_time is before the start_time? Shouldn't that be unexpected and worth raising an error?

Comment (naved001) [reply bare_metal_billing/models.py:38]: also, what if expire_time == start_time?

File: bare_metal_billing/main.py [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+
+    invoice_writer = billing.InvoiceWriter(
+        args.invoice_month,
+        project_invoices,
+        su_rates,
+        args.output_file,
+    )
+    invoice_writer.write_csv()
+
+    return 0
```
Comment (naved001) [inline bare_metal_billing/main.py:144]: why?

File: bare_metal_billing/billing.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    bm_usage_data: models.BMUsageData, start_time: datetime, end_time: datetime
+) -> list[models.ProjectUsage]:
+    project_usage_dict = {}
+    for lease_info in bm_usage_data.root:
+        project_name = lease_info.project
+        project_usage_dict.setdefault(
+            project_name, models.ProjectUsage(project_name=project_name, su_hours={})
+        )
+
+        su_type = _get_su_type(lease_info)
```
Comment (naved001) [inline bare_metal_billing/billing.py:113]: do you want to log if su_type is unkown kind? so when we run this we'll see it

File: bare_metal_billing/models.py [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+PositiveDecimalField = Annotated[Decimal, pydantic.Field(gt=0)]
+PositiveIntField = Annotated[int, pydantic.Field(gt=-1)]
+DateField = Annotated[datetime.datetime, pydantic.BeforeValidator(parse_date)]
+
+
+class BMNodeUsage(pydantic.BaseModel):
+    resource_class: Annotated[str, pydantic.Field(alias="Resource Class")]
+    project: Annotated[str, pydantic.Field(alias="Project")]
+    start_time: Annotated[DateField, pydantic.Field(alias="Start Time")]
+    expire_time: Annotated[DateField | None, pydantic.Field(alias="Expire Time")] = None
```
Comment (naved001) [inline bare_metal_billing/models.py:31]: I thought I commented it elsewhere, but it would be nice if we include the `resource` as well so we know exactly which node we billed for. Additionally, tracking the lease UUID will also be helpful. These may not be immediately useful when generating the billing/usage report, but it would be useuful information to have if we wanted to generate a report based on usage by each node.

File: bare_metal_billing/billing.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        time = max_time
+    return time
+
+
+def get_project_invoices(
+    bm_usage_data: models.BMUsageData, start_time: datetime, end_time: datetime
+) -> list[models.ProjectUsage]:
+    project_usage_dict = {}
+    for lease_info in bm_usage_data.root:
+        project_name = lease_info.project
```
Comment (naved001) [inline bare_metal_billing/billing.py:113]: Looking at the real data, I noticed that some of the leases had no project_name. Our model for BMUsageData allows for empty string as project name. I think for now that's okay until we figure out what's going on, but I'd like to log an error here that project_name is empty. In the report I generated for August, 2448 hours of FC430 was attributed to empty string project which adds up to $1836!

@tzumainn here's some of the leases that have no project. 
```json
  {
    "UUID": "0bed6c04-8a32-408c-99c0-2b54b20e8e69",
    "Resource": "MOC-R4PAC22U31-S3A",
    "Resource Class": "fc430",
    "Project": "",
    "Start Time": "2025-04-16T14:00:00",
    "End Time": "9999-12-31T00:00:00",
    "Offer UUID": null,
    "Status": "deleted",
    "Purpose": null
  },
  {
    "UUID": "f572ec82-7e69-4931-a186-af36a47257cf",
    "Resource": "MOC-R4PAC10U31-S3D",
    "Resource Class": "fc430",
    "Project": "",
    "Start Time": "2025-04-16T14:00:00",
    "End Time": "9999-12-31T00:00:00",
    "Offer UUID": null,
    "Status": "deleted",
    "Purpose": null
  },
  {
    "UUID": "eabfde7f-a940-4f07-920e-c8fcb8056d83",
    "Resource": "MOC-R4PAC08U37-S1D",
    "Resource Class": "fc430",
    "Project": "",
    "Start Time": "2025-04-16T14:00:00",
    "End Time": "9999-12-31T00:00:00",
    "Offer UUID": null,
    "Status": "deleted",
    "Purpose": null
  },
  ```

**Review by naved001 — COMMENTED** [review]
I generated a report for august, ran into some trouble so left the review. When I ran it I got warnings about unknown resource classes, so here are some more of them we need to track (in a different PR):

```
fc630-nvme
fc830-nvme
r620
r630
r720
r720xd
r730
r730xd
r740xd_amdmi210
whitebox-flax-1
whitebox-flax-2
```
@joachimweyl can tell us what SU types they should map to

File: bare_metal_billing/billing.py (hunk 22 above)
Comment (joachimweyl) [reply bare_metal_billing/billing.py:113]: For now, all those extra SU types are zero since we don't charge for them but there is the possibility we will in the future so make them individual SU types with 0 cost.

File: bare_metal_billing/models.py (hunk 21 above)
Comment (QuanMPhm) [reply bare_metal_billing/models.py:31]: Running the script myself I realized it would also be useful for logging

**Comment by QuanMPhm** [comment]
@joachimweyl On the BM invoice, do you want these extra SU types to appear as `BM Unknown` or as their names verbatim? I am wondering if that would be more informative for you when reviewing invoices

**Comment by joachimweyl** [comment]
names verbatim, makes sense now that we have a list of them.

**Comment by QuanMPhm** [comment]
@naved001 I've addressed your comments and added a bit more details to the logger messages

File: bare_metal_billing/billing.py (hunk 22 above)
Comment (tzumainn) [reply bare_metal_billing/billing.py:116]: This.... was an error on my part where, when cleaning up a project, I deleted the project itself, which resulted in leases losing knowledge of their project (that's why it's only `deleted` leases where you see this issue). I think at the time I told Kim, and he manually fixed some reports; it should not be an issue going forward.

**Review by larsks — APPROVED** [review]
There's a lot of data hardcoded into the scripts, which I think is going to make maintenance more difficult, but I don't think it's worth the effort to fix right now. Moving forward, we might want to consider moving all the data (su names, the resource-class-to-su mapping, etc) into files, and have the python code just implement the logic.
